diff -urNp -X bst-nodiff Bio-SamTools-1.33-orig/lib/Bio/DB/Bam/Alignment.pm Bio-SamTools-1.33/lib/Bio/DB/Bam/Alignment.pm
--- Bio-SamTools-1.33-orig/lib/Bio/DB/Bam/Alignment.pm	2011-12-06 21:15:01.000000000 +0000
+++ Bio-SamTools-1.33/lib/Bio/DB/Bam/Alignment.pm	2012-08-06 14:08:53.625662000 +0100
@@ -81,6 +81,26 @@ coordinates.
 
 Return the length of the alignment on the reference sequence.
 
+=item $mseqid = $align->mate_seq_id
+
+Return the seq_id of the mate's reference (target) sequence. This method
+is only available in the Bio::DB::AlignWrapper extension.
+
+=item $mstart = $align->mate_start
+
+For paired reads, return the start of the mate's alignment in
+1-based reference sequence coordinates.
+
+=item $mend = $align->mate_end
+
+For paired reads, return the end position of the mate's alignment in
+1-based reference sequence coordinates.
+
+=item $mlen = $align->mate_len
+
+For mate-pairs, retrieve the length of the mate's alignment on the
+reference sequence. 
+
 =item $strand = $align->strand
 
 Return the strand of the alignment as -1 for reversed, +1 for
@@ -294,6 +314,15 @@ The quality score for the alignment as a
 
 The bitwise flag field (see the SAM documentation).
 
+=item $mtid = $align->mtid
+
+For paired reads, the target ID of the mate's alignemnt.
+
+=item $mpos = $align->mpos
+
+For paired reads, the 0-based leftmost coordinate of the mate's
+alignment on the reference sequence.
+
 =item $n_cigar = $align->n_cigar
 
 Number of CIGAR operations in this alignment.
@@ -380,26 +409,6 @@ aligning.
 Return true if the aligned read's mate was reverse complemented prior
 to aligning.
 
-=item $mseqid  = $align->mate_seq_id
-
-Return the seqid of the mate.
-
-=item $mstart  = $align->mate_start
-
-For paired reads, return the start of the mate's alignment in
-reference sequence coordinates.
-
-
-=item $mend  = $align->mate_end
-
-For paired reads, return the end position of the mate's alignment. in
-reference sequence coordinates.
-
--item $len   = $align->mate_len
-
-For mate-pairs, retrieve the length of the mate's alignment on the
-reference sequence. 
-
 =item $isize = $align->isize
 
 For mate-pairs, return the computed insert size.
@@ -466,7 +475,7 @@ sub get_all_tags {
 
 sub start {
     my $self = shift;
-    return if $self->unmapped;
+    return if $self->pos < 0 || $self->unmapped;
     return $self->pos+1;
 }
 
@@ -558,7 +567,9 @@ sub length {
 }
 
 sub mate_start {
-    shift->mpos+1;
+    my $self = shift;
+    return if $self->mpos < 0 || $self->munmapped;
+    return $self->mpos+1;
 }
 
 sub mate_len {
diff -urNp -X bst-nodiff Bio-SamTools-1.33-orig/lib/Bio/DB/Bam/AlignWrapper.pm Bio-SamTools-1.33/lib/Bio/DB/Bam/AlignWrapper.pm
--- Bio-SamTools-1.33-orig/lib/Bio/DB/Bam/AlignWrapper.pm	2011-12-06 21:15:01.000000000 +0000
+++ Bio-SamTools-1.33/lib/Bio/DB/Bam/AlignWrapper.pm	2012-08-06 11:31:27.477164000 +0100
@@ -23,7 +23,9 @@ L<Bio::DB::Bam::Alignment/High-level Bio
  seq_id()              return human-readable reference sequence name
  seq()                 return Bio::PrimarySeq object for reference sequence
  subseq()              return a subsequence across the indicated range
+ mate_seq_id()         return human-readable mate reference sequence name
  dna()                 return the DNA of the reference sequence
+ tam_line()            return the text representation of the alignment
  attributes()          synonym for get_tag_values()
  get_all_tags()        return all tag names
  get_tag_values()      return the values of the given tag
@@ -174,13 +176,14 @@ sub hit { shift->target(@_); }
 sub seq_id {
     my $self = shift;
     my $tid  = $self->tid;
+    return if $tid < 0 || $self->unmapped;
     $self->{sam}->target_name($tid);
 }
 
 sub mate_seq_id {
     my $self = shift;
     my $tid  = $self->mtid;
-    return unless $tid >= 0;
+    return if $tid < 0 || $self->munmapped;
     $self->{sam}->target_name($tid);
 }
 
@@ -403,16 +406,16 @@ sub tam_line {
     return join ("\t",
 		 $self->qname,
 		 $self->flag,
-		 $self->seq_id,
+		 $self->tid >= 0 ? $self->{sam}->target_name($self->tid) : '*',
 		 $self->pos+1,
 		 $self->qual,
-		 $self->cigar_str,
-		 $self->mate_seq_id ? ($self->mate_seq_id eq $self->seq_id ? '=' : $self->mate_seq_id) : '*',
-		 $self->mpos || 0,
+		 $self->cigar_str || '*',
+		 $self->mtid >= 0 ? ($self->mtid == $self->tid ? '=' : $self->{sam}->target_name($self->mtid)) : '*',
+		 $self->mpos+1,
 		 $self->isize,
 		 $self->qseq,
 		 join('',map{chr($_+33)} $self->qscore),
-		 $self->aux
+		 $self->aux || ()
 	);
 }
 
diff -urNp -X bst-nodiff Bio-SamTools-1.33-orig/lib/Bio/DB/Sam.xs Bio-SamTools-1.33/lib/Bio/DB/Sam.xs
--- Bio-SamTools-1.33-orig/lib/Bio/DB/Sam.xs	2011-12-06 21:15:01.000000000 +0000
+++ Bio-SamTools-1.33/lib/Bio/DB/Sam.xs	2012-07-31 17:01:06.924274000 +0100
@@ -551,8 +551,6 @@ PROTOTYPE: $;$
 CODE:
     if (items > 1)
       b->core.mpos = SvIV(ST(1));
-    if (b->core.pos <= 0)
-      XSRETURN_UNDEF;
     RETVAL=b->core.mpos;
 OUTPUT:
     RETVAL
